edgelet/edgelet-core/src/watchdog.rs

```diff
@@ -110,22 +116,41 @@ where
         "Starting watchdog with {} second frequency...",
         WATCHDOG_FREQUENCY_SECS
     );
+
     Interval::new(Instant::now(), Duration::from_secs(WATCHDOG_FREQUENCY_SECS))
         .map_err(|err| Error::from(err.context(ErrorKind::EdgeRuntimeStatusCheckerTimer)))
-        .for_each(move |_| {
+        .and_then(move |_| {
             info!("Checking edge runtime status");
             check_runtime(
                 runtime.clone(),
                 id_mgr.clone(),
                 spec.clone(),
                 module_id.clone(),
             )
+            .and_then(|_| future::ok(None))
             .or_else(|e| {
                 warn!("Error in watchdog when checking for edge runtime status:");
                 log_failure(Level::Warn, &e);
-                future::ok(())
+                future::ok(Some(e))
             })
         })
+        .fold(
+            0,
+            move |exec_count: u32, result: Option<Error>| match result {
```

Comment:
with the retry limit enum you can do:

```rust
move |exec_count: u32, result: Option<Error>| {
    result.and_then(|e| {
        if retry_count.cmp(exec_count) == Ordering::Greater {
            Ok(exec_count + 1)
        } else {
            Err(e)
        }
    })
    .unwrap_or_else(|| Ok(0))
}
```

Using a `match` with `Option` is generally seen as an anti-pattern. Especially, when there are only two branches, a `Some` and `None`. People tend to use `map`/`and_then` paired with one of the `unwrap_` methods.